File: src/de/blackpinguin/android/sindwirschonda/si/SIValue.java
package de.blackpinguin.android.sindwirschonda.si;

public class SIValue {
	
	
	
	/*
	 * Subklasse
	 */
	public static class NotAValue extends SIValue{
		private NotAValue(){
			super(Double.NaN, SIUnit.NaU);
		}
	}
	
	public static final NotAValue NaV = new NotAValue();
	
	
	
	/*
	 * Instanzvariablen
	 */
	private final double value;
	private final SIUnit unit;
	
	public double getValue(){return value;}
	public SIUnit getUnit(){return unit;}
	
	
	/*
	 * Konstruktor
	 */
	public SIValue(double value, SIUnit unit){
		this.value = value;
		this.unit = unit;
	}
	
	
	
	/*
	 * Konvertierung
	 */
	private SIValue _asBaseUnit = null;
	public SIValue toBaseUnit(){
		if(_asBaseUnit == null)
			if(unit == unit.getBaseUnit())
				_asBaseUnit = this;
			else
				_asBaseUnit = new SIValue(unit.toBaseUnit(value).getValue(), unit.getBaseUnit()); 
		return _asBaseUnit;
	}
	
	public SIValue toUnit(SIUnit newUnit){
		if(this.getUnit().getBaseUnit() != newUnit.getBaseUnit())
			return NaV;
		return newUnit.fromBaseUnit(this.toBaseUnit().getValue());
	}
	
	
	
	/*
	 * Operatoren
	 */
	public SIValue add(SIValue other){
		if(this.getUnit().getBaseUnit() != other.getUnit().getBaseUnit())
			return NaV;
		
		double result = this.toBaseUnit().getValue() + other.toBaseUnit().getValue();
		
		return this.getUnit().fromBaseUnit(result);
	}
	
	public SIValue multiply(SIValue other){
		SIUnit bunit = this.getUnit().multiply(other.getUnit());
		if(bunit == SIUnit.NaU)
			return NaV;
		return new SIValue(this.toBaseUnit().getValue() * other.toBaseUnit().getValue(), bunit);
	}
	
	public SIValue divide(SIValue other){
		SIUnit bunit = this.getUnit().divide(other.getUnit());
		if(bunit == SIUnit.NaU)
			return NaV;
		return new SIValue(this.toBaseUnit().getValue() / other.toBaseUnit().getValue(), bunit);
	}
	
	
	
	
	
}

File: src/de/blackpinguin/android/sindwirschonda/si/SIValue.scala
package de.blackpinguin.android.sindwirschonda.si

//Begleitobjekt ("static")
object SIValue {

  //SIUnit lässt sich nicht einfach serialisieren, deshalb manuelle Serialisierung
  case class Serialized(value: Double, unitType: String, unit: String) {
    def unserialize: SIValue = SIValue(value, SIUnitType(unitType)(unit))
  }

  object NotAValue extends SIValue(Double.NaN, SIUnit.NaU)

  val NaV = NotAValue
}

//Klasse
case class SIValue(value: Double, unit: SIUnit) {

  //SIUnit lässt sich nicht einfach serialisieren, deshalb manuelle Serialisierung
  def serialize = SIValue.Serialized(value, unit.unitType.name, unit.abbreviation)

  //Dieses Value-Objekt umwandeln in die Basiseinheit
  lazy val toBaseUnit =
    if (unit == unit.getBaseUnit) this
    else SIValue(unit.toBaseUnit(value), unit.getBaseUnit)

  //diese Größe umrechnen in eine andere Einheit
  def toUnit(newUnit: SIUnit) =
    if (unit.getBaseUnit != newUnit.getBaseUnit)
      SIValue.NaV
    else
      newUnit.fromBaseUnit(this.toBaseUnit)

  /*
   * Operationen
   */

  //Addition
  def +(other: SIValue): SIValue =
    //nicht identische Größen
    if (unit.getBaseUnit != other.unit.getBaseUnit)
      SIValue.NaV
    //identische Größen
    else
      unit.fromBaseUnit(this.toBaseUnit.value + other.toBaseUnit.value)

  //Subtraktion
  def -(other: SIValue): SIValue =
    //nicht identische Größen
    if (unit.getBaseUnit != other.unit.getBaseUnit)
      SIValue.NaV
    //identische Größen
    else
      unit.fromBaseUnit(this.toBaseUnit.value - other.toBaseUnit.value)

  //Multiplikation
  def *(other: SIValue): SIValue = {
    //Einheit die bei der Operation resultiert
    val u = unit * other.unit
    //keine bekannte Ausgabeeinheit
    if (u == SIUnit.NaU)
      SIValue.NaV
    //Operation ist möglich
    else
      SIValue(toBaseUnit.value * other.toBaseUnit.value, u)
  }

  //Division
  def /(other: SIValue): SIValue = {
    //Einheit die bei der Operation resultiert
    val u = unit / other.unit
    //keine bekannte Ausgabeeinheit
    if (u == SIUnit.NaU)
      SIValue.NaV
    //Operation ist möglich
    else
      SIValue(toBaseUnit.value / other.toBaseUnit.value, u)
  }

  def *(scale: Double): SIValue = SIValue(value * scale, unit)
  def /(scale: Double): SIValue = SIValue(value / scale, unit)

  //andere Methodennamen für Java-Anwendungen
  def add(other: SIValue) = this + other
  def sub(other: SIValue) = this - other
  def mult(other: SIValue) = this * other
  def div(other: SIValue) = this / other
  def mult(scale: Double) = this * scale
  def div(scale: Double) = this / scale

  override def toString = value.toString + " " + unit.abbreviation
}